src/t_stream.c

```diff
@@ -1809,20 +1824,136 @@ size_t streamReplyWithRange(client *c, stream *s, streamID *start, streamID *end
+
+            streamID pel_id;
+            streamIteratorStart(&si,s,&pelKey->id,&pelKey->id,rev);
+            if (streamIteratorGetID(&si,&pel_id,&numfields)) {
+                /* Emit a four elements array: ID, array of field-value pairs,
+                 * idle time and delivery count. */
+                addReplyArrayLen(c,4);
+                addReplyStreamID(c,&pel_id);
+                addReplyArrayLen(c,numfields*2);
+
+                /* Emit the field-value pairs. */
+                while (numfields--) {
+                    unsigned char *key, *value;
+                    int64_t key_len, value_len;
+                    streamIteratorGetField(&si,&key,&value,&key_len,&value_len);
+                    addReplyBulkCBuffer(c,key,key_len);
+                    addReplyBulkCBuffer(c,value,value_len);
+                }
+
+                uint64_t idle = commandTimeSnapshot() - pelKey->delivery_time;
+                addReplyBulkLongLong(c, idle);
+                addReplyBulkLongLong(c, delivery_count);
+
+                /* Remove the NACK from old consumer and time-based PEL. */
+                raxRemove(nack->consumer->pel,buf,sizeof(buf),NULL);
+                raxRemovePelByTime(group->pel_by_time, nack->delivery_time, &pel_id);
+
+                /* Transfer NACK to new consumer with updated metadata. */
+                nack->consumer = consumer;
+                nack->delivery_time = commandTimeSnapshot();
+                nack->delivery_count++;
+                raxInsert(consumer->pel,buf,sizeof(buf),nack,NULL);
+                raxInsertPelByTime(group->pel_by_time, nack->delivery_time, &pel_id);
+
+                consumer->active_time = commandTimeSnapshot();
+
+                /* Propagate as XCLAIM. */
+                if (spi) {
+                    robj *idarg = createObjectFromStreamID(&pel_id);
+                    streamPropagateXCLAIM(c,spi->keyname,group,spi->groupname,idarg,nack);
+                    decrRefCount(idarg);
+                    if (propCount) (*propCount)++;
+                }
+
+                arraylen++;
+            }
+            streamIteratorStop(&si);   
+        }
+        listRelease(eligible_pels);
+    }
     /* If the client is asking for some history, we serve it using a
      * different function, so that we return entries *solely* from its
      * own PEL. This ensures each consumer will always and only see
      * the history of messages delivered to it and not yet confirmed
      * as delivered. */
     if (group && (flags & STREAM_RWR_HISTORY)) {
         return streamReplyWithRangeFromConsumerPEL(c,s,start,end,count,
-                                                   consumer);
+                                                   group, consumer);
     }
 
-    if (!(flags & STREAM_RWR_RAWENTRIES))
+    /* Stop here if client only wants claimed entries or count is satisfied. */
+    if ((group && (flags & STREAM_RWR_CLAIMED)) || (count && count == arraylen)) {
```

Comment:
What if arraylen < count? At this point, arraylen_ptr has already replied. If we don't go into this if, will it cause the reply to be corrupted?

Reply:
I almost understand. The response from xreadgroup will include claimed enties and entries that haven't been read. Does our current test cover this? I'm not sure if the current schema definition will handle it correctly.